crate/operator/utils/kopf.py

```diff
@@ -26,3 +40,73 @@ async def _wrapper(**_):
         await awaitable(*args, **kwargs)
 
     return _wrapper
+
+
+class StateBasedSubHandler(abc.ABC):
+    """
+    State-based sub-handlers allow running sub-handlers in a defined order.
+
+    Kopf sub-handlers run in parallel, meaning, the order of them being
+    executed is not strictly defined. When updating a custom CrateDB resource,
+    though, the order of the steps to deal with the changes in the operator are
+    predefined.
+
+    Subclasses of :class:`StateBasedSubHandler` are required to implement the
+    :meth:`~StateBasedSubHandler.handle` method, which is the code that'll run
+    as a sub-handler. The class will ensure that the sub-handler is only run
+    when the current state in :class:`the statemachine
+    <crate.operator.utils.state.StateMachine>` is :attr:`state`.
+    """
+
+    #: The :class:`~crate.operator.utils.state.State` at which to run the
+    #: subhandler. Until the state has been reached, the sub-handler will be
+    #: retried using :class:`kopf.TemporaryError`.
+    state: State
+
+    def __init__(self, namespace: str, name: str, context: Context):
+        self.namespace = namespace
+        self.name = name
+        self._context = context
+
+    def __call__(self):
+        return functools.partial(self._subhandler)
```

Comment:
What difference does it make whether you return `self._subhandler` directly or as as partial without extra arguments?

Reply:
I couldn't get it to work with `return self._subhandler`. or doing what's happening in `_subhandler` in `__call__` directly. The issues I faced were around pickling some thing in the inner parts of Kopf, where I then didn't spent time investigating. I _suspect_ it has something to do with how `kopf.register()` and async handlers / sub-handlers work. But yeah, I'm not too sure what exactly the underlying issue is.